### ISSUE TITLE
Champion "blittable"

### ISSUE THREAD

**Issue by gafter** [dotnet/csharplang#187]
- [ ] Proposal added
- [ ] Discussed in LDM
- [ ] Decision in LDM
- [ ] Finalized (done, rejected, inactive)
- [ ] Spec'ed

**Comment by eyalsk** [comment]
Is this a proposal to improve existing blittable types or something? what's this about? :)

**Comment by vbcodec** [comment]
@eyalsk  It si [Redacted] ::)

**Comment by ufcpp** [comment]
Is this related to https://github.com/dotnet/roslyn/issues/2209 ?

**Comment by bbarry** [comment]
and/or [dotnet/roslyn#7592](https://github.com/dotnet/roslyn/issues/7592)? Or something to do with compile time evaluation for structs?

**Comment by jaredpar** [comment]
Sorry for the confusion.  Will reopen when we have the proposal ready.

**Comment by jaredpar** [comment]
Sorry there isn't more details here.  This is being used as a reminder for me to add more details.

**Comment by vbcodec** [comment]
verfy funny :)
Rename to 'Reservation place for any incoming proposal'

**Comment by jaredpar** [comment]
Proposal added here https://github.com/dotnet/csharplang/pull/206

**Comment by ufcpp** [comment]
This is good idea! but I have some questions

- How is this related to [Blittable Types in Interop Marshaling](https://msdn.microsoft.com/en-us/library/75dwhxf7)? The same? or similar but bit different?
- Can `IntPtr` and `UIntPtr` be blittable? They are blittable in the Interop Marshaling and unmanaged in the current C# spec.

**Comment by eyalsk** [comment]
@jaredpar Just a thought.

Why can't it be an attribute?

```
[Blittable]
struct Point 
{
    public int X;
    public int Y;
}
```

And instead of adding a new `blittable` constraint such in the following case:

```
void Hash<T>(T value) where T : blittable struct
{
    using (T* p = &value) { 
        ...
    }
}
```

Allow attributes to be used as a generic constraint? so something like this.

```
void Hash<T>(T value) where T : struct, Blittable
{
    using (T* p = &value) { 
        ...
    }
}
```

Just seems like a more general feature that can be used here and elsewhere in the framework. :)

**Comment by eyalsk** [comment]
@bbarry Why the downvote? what's the point of adding a new keyword? and then adding a special constraint when existing features can actually do the job or/and improved? not to mention that `blittable struct` is at odds with existing constraints, for example, why can't we check that the generic parameter is `abstract class`? what makes `blittable` any special?

Maybe I'm overlooking something so if you can, please do elaborate. :)

**Comment by KodrAus** [comment]
@eyalsk I think it's reasonable as a keyword, because adding it to a structure changes its compile-time behaviour (it means adding non-blittable fields will fail to compile). It seems like a nice fit to me.

**Comment by bbarry** [comment]
@eyalsk, it is nothing personal, my downvote is just that I'd rather see this be a keyword over an attribute. It is admittedly a fine line in the distinction (surely the compiled code would contain some attribute so that this can be enforced across assemblies, so the compiler must in fact know about that as it is enforced).

Adding a special case to permit the use of one particular attribute for this purpose seems arbitrary and counter to syntax consistency and rules that devs can internalize when writing code (I dislike all the attributes that do this already). Allowing general constraints in the form `where T : Something`  where `Something`  maps to an attribute type `SomethingAttribute` which must be defined on the type `T` is at first mildly interesting (I am not sure it has advantages over a marker interface, but I don't really like those), but is significantly a wider scope than this proposal and at best should be separate. 

My gut says the former is yucky and the latter is going to introduce either ambiguities or breaking changes. A keyword isn't significantly better, but I can reasonably expect that a keyword means something to a compiler and will have an impact on how my code builds. 

---

If there were the ability to have generalized attribute constraints, `Blittable` would still need to be handled in a special way by the compiler because it must enforce certain constraints on the `struct` type where it is used (for example disallowing non-blittable fields or auto struct layout).

**Comment by eyalsk** [comment]
@KodrAus, @bbarry Okay, fair enough, thank you both for the explanation. :)

**Comment by HaloFour** [comment]
The name `blittable` is a bit weird.  To newcomers it doesn't remotely convey what it means.  I'm not convinced that "blitting" is a known concept outside of a specific subset of Win32 developers who happened to know that "BitBlt" inspired a verb, and I'm one of them.

The name `unmanaged` is a little better, but it doesn't really grab me.  I don't know that I have a better option.  Maybe `fixed`, but that seems to imply more than this does as well.

Also, how is the generic constraint enforced?  I don't see a mention of a CLR change to accompany this, so would this be encoded on the generic type/method as an attribute and enforced only by supporting compilers?  What would happen if an errant compiler were to use one of these generic types/methods specifying a non-`blittable` struct generic type argument?

**Comment by ufcpp** [comment]
@HaloFour 

> What would happen if an errant compiler

This problem already exists in the current C# compiler. Using the [`Unsafe`](https://www.nuget.org/packages/System.Runtime.CompilerServices.Unsafe/) class, managed types can be converted to pointer types. This could cause GC crash.

```cs
using System;
using System.Runtime.CompilerServices;

struct UnmanagedStruct
{
    public int X;
    public int Y;
    public override string ToString() => (X, Y).ToString();
}

struct ManagedStruct
{
    public string X;
    public string Y;
    public override string ToString() => (X ?? "NULL", Y ?? "NULL").ToString();
}

unsafe class Program
{
    static void Main()
    {
        // OK
        // An unmanaged type can be used with pointers
        UnmanagedStruct us = new UnmanagedStruct { X = 1, Y = 2 };
        var usp = (int*)Unsafe.AsPointer(ref us);

        usp[0] = 100;
        usp[1] = 200;
        Console.WriteLine(us); // (100, 200)

        // OK!
        // This is an abuse of the Unsafe class.
        // The Unsafe class is implemented by IL and allows some problematic codes.
        // For instance, it can make managed type pointer.
        ManagedStruct ms = new ManagedStruct { X = "abc", Y = "cde" };
        var msp = (IntPtr*)Unsafe.AsPointer(ref ms);

        msp[0] = (IntPtr)0;
        msp[1] = (IntPtr)0;
        Console.WriteLine(ms); // (NULL, NULL)

        msp[0] = (IntPtr)1234567; // invalid pointer. GC could be crashed
    }
}
```

**Comment by ufcpp** [comment]
Related to https://github.com/dotnet/corefxlab/blob/e109353256875ee42e6365c5facd243b1f0497ce/src/System.Slices/System/SpanExtensions_binary.cs#L21

**Comment by HaloFour** [comment]
@ufcpp 

Yes, but the existing code doesn't have any declared "constraints" intended to prevent that so the developer is definitely walking into unsafe territory there.

**Comment by jaredpar** [comment]
@eyalsk 

> Why can't it be an attribute?

In general the C# language avoids attaching semantics to attributes attached to types / methods.  The preference is to use explicit language syntax to implement semantic changes in behavior.

**Comment by eyalsk** [comment]
@jaredpar Understood, thanks. :)

**Comment by jaredpar** [comment]
@HaloFour 

>The name blittable is a bit weird.
> The name unmanaged is a little better, but it doesn't really grab me. 

Yep ... naming is hard.  😄 

One of the other names we considered in Midori was `primitive`.

**Comment by jnm2** [comment]
I like `blittable`. My estimate of developers would be more liberal, but it's easy enough to Google if you are unfamiliar with a term. The win is that `blittable` is the most directly precise term.

**Comment by HaloFour** [comment]
@jaredpar 

Agreed, and it's a minor point at least at this stage of the proposal.

@jnm2 

Even if that were the case, `blit` by itself doesn't mean anything.  It happens to be the pseudo-pronouciation given to 1970s-era graphics compositing routines that Win32 happened to inherit but are largely obsolete today.  We might as well call them sprite structs.  I'd rather a name that better captures the intent of constraining the type to primitives or composites of primitives.

**Comment by jnm2** [comment]
I've often seen "blit" used without a graphics connotation, more of a "spit out these raw bytes fast without parsing" connotation which is even slightly contradictory to the original meaning.

**Comment by orthoxerox** [comment]
Yeah, 'blittable' means 'can be safely treated as a single contiguous chunk of bits' to me.

**Comment by mikedn** [comment]
> How is this related to Blittable Types in Interop Marshaling? The same? or similar but bit different?
>> It's the same

@jaredpar Almost but not quite the same. Unfortunately `bool` isn't blittable according to marshaling but it is according to this proposal.

**Comment by omariom** [comment]
`flat`?

**Comment by nietras** [comment]
'plain' perhaps e. g. 'plain struct' or 'pure', but 'blittable' doesn't work we've had that discussion many times before. I think I made a list of possible names at some point, can't find it though. 😕

**Comment by eyalsk** [comment]
I don't know but maybe `interoperable struct`? yes it's a bit long but at least people would understand what it is or at least where it should be used? alternatively `interop struct`?

**Comment by Thaina** [comment]
I was propose #144 that, in my opinion, struct should always automatically inherit the intersection of the kind of any fields it contains

If any struct contain only blittable, then it should also be blittable

could we have compiler automatically attach attribute or derived from difference type than ValueType (make Primitive type, Integer type, Number type, Blittable type, etc. derived from ValueType?)

**Comment by tannergooding** [comment]
Blittable is used in a large majority of the existing MSDN documentation (example: https://msdn.microsoft.com/en-us/library/75dwhxf7(v=vs.110).aspx), so I think it makes the most sense. Anyone who does large amounts of Interop (or who has read the CLR Interop documentation) should be familiar with the term.

Also, `bool`, `char`, and `string` are special. They are considered "sometimes blittable" and have special handling in the runtime. For example, `char` is blittable when the type is 2-bytes on both sides (sometimes `char` needs to be 1 or 4 bytes, however) and `bool` is blittable when the type is 1-byte on both side (most of the time, for Windows interop, `bool` needs to be 4-bytes, however). In certain cases, `string` will also be pinned and marshaled directly as well (rather than copied and then marshaled).

**Comment by aluanhaddad** [comment]
This sounds like a great feature but I agree with @HaloFour that the name is far from ideal. That isn't a reason to hold back implementing it and I like the idea of using `fixed`.

I could be completely wrong but from what I understand, the meaning is something like value type composed only of value types. That is, I think it means something is a `deep` value type. Is my understanding correct?

If it is, perhaps `independent`, `internallycontiguous`, or something that sounds better than that but conveys the idea could work better.

**Comment by tannergooding** [comment]
@aluanhaddad, you are correct. It is a value type that is composed purely of other value types. That is, there should be no reference types in the hierarchy: no strings, no arrays, no objects, etc...

One of the caveats, however, is that there are a few classes of `value type` objects which may not be allowed (they have special marshaling semantics). This set includes `bool` and `char`.

I do agree that the term `blittable` can be a bit weird, but (as per my previous post), the vast majority of existing .NET Framework documentation on the subject refers to this as "blittable" (a type which can be 'bit blitted', or copied without conversion, when sent between managed/unmanaged code or from one process to another).

However, anyone remotely familiar with interop code or the marshalling behaviors of the runtime should be familiar with the term (as it is likely they read some of the MSDN documentation on the subject to understand the behaviors).

I don't think that `independent` or `internallycontiguous` conveys the correct meaning here. The only other terms suggested so far have been:
* `fixed`: This has an existing meaning and is only useable in unsafe code, it would likely be the most confusing
* `flat`: This could be implying a number of other things and does not correctly relay the fact that this can be structured data (pointers are blittable as well)
* `unmanaged`: This one makes sense for the unmanaged/managed case but not so much for the IPC case
* `interop`: This one also makes sense, but it is not a widely used existing term used to refer to these types

Additionally, as @jnm2 pointed out, googling these terms also shows which are the most relevant. There are a number of terms that come up for each one and `blittable` is the only one that has correct and relevant links (on both Bing and Google). 

At least for me:
 * The MSDN article on `Blittable and Non-Blittable Types` is the top result: https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types
 * The Wikipedia article on `Blittable types` is the second result: https://en.wikipedia.org/wiki/Blittable_types
 * There are various blog posts and SO articles/questions on the subject following that in the search results

**Comment by Joe4evr** [comment]
> One of the caveats, however, is that there are a few classes of value type objects which may not be allowed (they have special marshaling semantics). This set includes `bool` and `char`.

Also `Nullable<T>` (and other generic structs). 

But yeah, naming things and cache invalidation, right? 🙃

**Comment by tannergooding** [comment]
> Also Nullable<T> (and other generic structs).

Generic structs probably depends on whether or not we also get a `blittable `constraint, but yes 😄 

> But yeah, naming things and cache invalidation, right?
I'm still convinced naming things is the harder problem :trollface:

**Comment by nietras** [comment]
I think the point here should be there are two different scenarios; interop
and unmanaged memory. blittable is for interop, but we really need
something for `plain` or `pure` value types that holds for both char and
bool as well. That's why blittable should only be used for marshalling
scenarios.

**Comment by alrz** [comment]
>I'm still convinced naming things is the harder problem

Also, [version mapping](https://github.com/dotnet/standard/blob/master/docs/versions.md) ☠️ .

**Comment by jnm2** [comment]
@alrz Does that fall under "off-by-one errors?"

**Comment by aluanhaddad** [comment]
@tannergooding
I wasn't actually arguing for using thosr names, just throwing out strawmen.

**Comment by MarkPflug** [comment]
Several people have expressed displeasure with the `blittable` keyword, and I think I agree; it feels weird for some reason. What about `value struct`? `value` is already a contextual keyword, can it be reused in this context? `value` feels pretty accurate to me, in that it is a struct that contains no "refs". Maybe I'm wrong about my understanding of this though. Or possibly `value` wouldn't work because it would cause ambiguity in the language somewhere.

**Comment by alrz** [comment]
[static delegates](https://github.com/dotnet/csharplang/blob/master/proposals/static-delegates.md) are blittable,
```cs
static delegate int Func()
```
So I'd say `static struct` is consistent in that manner, at least.

**Comment by tannergooding** [comment]
Some structs that contain references can still be blittable (strings, once pinned, and assuming they are passed as readonly are blittable, for example).

I still think that `blittable` matches all the existing documentation on the subject and is easily found if unfamiliar (top result for "blittable" on both Bing and Google).

I think that most of the other suggestions come close, but miss scenarios that would otherwise be ok.

**Comment by alrz** [comment]
`blittable` has the same problem that `supersede` modifier had: looks like we took it from Eminem's lyrics.

Seriously though, @tannergooding, do you think `static struct` vs `static delegate` doesn't make sense neither? Regarding documentations on the subject, we did this with `struct` and `class` for value vs. reference types. I don't think we should reflect the exact wording of the documentation, rather we should strive for consistency in the language, IMO.

**Comment by jnm2** [comment]
`public static class Foo` => `public static struct Foo` => wheee!

**Comment by MarkPflug** [comment]
> Some structs that contain references can still be blittable (strings, once pinned, and assuming they are passed as readonly are blittable, for example).

Is that not true for any pinned reference, a pinned byte[] array? Or are strings truly special in this regard? Wouldn't this just be a effectively a void* or a IntPtr then? Still sounds "value-ish" to me.

**Comment by jnm2** [comment]
> Wouldn't this just be a effectively a void* or a IntPtr then? Still sounds "value-ish" to me.

To me, value-ish, no. Blit-ish, yes. For what it's worth. 😄 

I think blit has more apropos meaning in people's minds (“this thing can be used without transformation”) compared to value (“this thing has value semantics rather than identity semantics”). or even `val` which some propose to mean `readonly var`.

**Comment by tannergooding** [comment]
@alzr, I'm not convinced `static delegate` is right either. I had initially proposed `value delegate`, but some discussion ended up changing that (I don't recall what it was at the time).

Code that you aren't familiar with is going to be searched and `static struct` will likely cause ambiguity and confusion. I think `static struct` will cause large amounts of confusion, especially when compared to `static class`.

**Comment by tannergooding** [comment]
@MarkPflug, not all pinned references are blittable. Strings are sometimes blittable, single dimensional 0-based arrays of blittable types are themselves blittable:
https://docs.microsoft.com/en-us/dotnet/framework/interop/blittable-and-non-blittable-types
https://github.com/dotnet/corefx/blob/04b84e0206bbb3d126fdc3329b6417971d3bc895/Documentation/coding-guidelines/interop-pinvokes.md
https://github.com/dotnet/corefx/blob/42ce0e696e08e190aceaf8613e47e67d57178550/Documentation/coding-guidelines/interop-guidelines.md

**Comment by nietras** [comment]
Should the title of this not be renamed to lose the "blittable" naming now that the wording has, thank god, become `unmanaged`? 😄

**Comment by OmarTawfik** [comment]
Assigning back to champion since implementation is done.

**Comment by xoofx** [comment]
Taking a pointer to a generic with unmanaged constraint (or indirectly on a struct that is implicitly unmanaged transitively as for issue #1504) is the very primary usage of the unmanaged constraint. I can only consider the `unmanaged` constraint done if `T*` and `fixed` on T are allowed at C# level. This is possible at the IL level, this should be unlocked at C# level so that we can get rid off once and for good of the IL tricks we have been using for years to workaround this problem. It should allow to remove a dependency to `System.Runtime.CompilerServices.Unsafe`.
Is there any thing blocking this at the language spec and furthermore at the roslyn level?

**Comment by xoofx** [comment]
Oops, nevermind, retrying this with latest and it seems that pointers/fixed as supported. The only remaining problem is #1504 which can be very annoying but less critical

**Comment by yaakov-h** [comment]
> Blittable gives me the guarantee that I can take a pointer to it, read it, write it to a stream and recover it somewhere else without losing the meaning.

Assuming same CPU bitness and endian-ness?

> I do not think that is true of char, I should be using a character encoding.

System.Char is UTF-16, IIRC.
